processing/src/main/java/io/druid/query/aggregation/cardinality/CardinalityAggregatorFactory.java

```diff
@@ -35,17 +35,49 @@
 import io.druid.query.aggregation.hyperloglog.HyperLogLogCollector;
 import io.druid.query.aggregation.hyperloglog.HyperUniquesAggregatorFactory;
 import io.druid.query.dimension.DefaultDimensionSpec;
+import io.druid.query.dimension.DimensionSpec;
 import io.druid.segment.ColumnSelectorFactory;
 import io.druid.segment.DimensionSelector;
 import org.apache.commons.codec.binary.Base64;
 
-import javax.annotation.Nullable;
 import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
 
 public class CardinalityAggregatorFactory extends AggregatorFactory
 {
+  private static List<String> makeFieldNamesFromFields(List<DimensionSpec> fields)
+  {
+    return Lists.transform(
+        fields,
+        new Function<DimensionSpec, String>()
+        {
+          @Override
+          public String apply(DimensionSpec input)
+          {
+            return input.getOutputName();
+          }
+        }
+    );
+  }
+
+  private static List<DimensionSpec> makeFieldsFromFieldNames(List<String> fieldNames)
+  {
+    return Lists.transform(
```

Comment:
Lists.transform is lazy; since this is going into a member field, would probably be better to copy by wrapping this into `ImmutableList.copyOf(...)`